### PR TITLE
Provide proper compilation errors for invalid interop usages

### LINKED ISSUE #40327: [Improvement]: Provide compile-time errors for invalid interop assignments
### Description

Need to provide a compile-time errors for the invalid interop assignments such as the below case such that it will not crash at runtime.
```bal

import ballerina/jballerina.java;

public function main() {
    error e = error("err");
    handle printableStackTrace = getPrintableStackTrace(e);
}

function getPrintableStackTrace(error bError) returns handle = @java:Method {
    'class: "io.ballerina.runtime.api.values.BError"
} external;
```
gives
```
Compiling source
        issue.bal
WARNING [issue.bal:(6:5,6:60)] unused variable 'printableStackTrace'

Running executable

[2023-04-28 12:50:28,526] SEVERE {b7a.log.crash} - Bad type on operand stack
Exception Details:
  Location:
    issue.getPrintableStackTrace(Lio/ballerina/runtime/internal/scheduling/Strand;Lio/ballerina/runtime/internal/values/ErrorValue;)Lio/ballerina/runtime/internal/values/HandleValue; @144: invokevirtual
  Reason:
    Type 'io/ballerina/runtime/internal/values/ErrorValue' (current frame, stack[0]) is not assignable to 'io/ballerina/runtime/internal/values/HandleValue'
  Current Frame:
    bci: @144
    flags: { }
    locals: { 'io/ballerina/runtime/internal/scheduling/Strand', 'io/ballerina/runtime/internal/values/ErrorValue', 'java/lang/String', 'io/ballerina/runtime/internal/values/HandleValue', integer, null, null, integer }
    stack: { 'io/ballerina/runtime/internal/values/ErrorValue' }
  Bytecode:
    0000000: 014d 014e 0336 0401 3a05 013a 0603 3607
    0000010: 2ab4 0011 9d00 e02a 59b4 0014 5a04 60b5
    0000020: 0014 3607 2ab4 0018 9900 07b8 001e bf15
    0000030: 04ab 0000 0000 00f4 0000 0004 0000 0000
    0000040: 0000 002b 0000 0001 0000 0030 0000 0002
    0000050: 0000 00a6 0000 0003 0000 00bc 1100 0036
    0000060: 0411 0001 3604 2ab6 002c 9900 252a 03b5
    0000070: 002f 2ab4 0033 c600 0e2a b400 3359 2a01
    0000080: b500 33bf 2ab4 010d c001 0a4d a700 202b
    0000090: b601 10c0 00b7 59c7 0011 b201 15b2 011b
    00000a0: 03bd 0004 b801 21bf b601 244d 2ab6 005c
    00000b0: 9900 0f13 0126 3a05 1260 3a06 a700 69a7
    00000c0: 0003 2c59 c101 289a 0006 b801 2cb2 0130
    00000d0: b801 36c0 0106 4e11 0002 3604 a700 1100
    00000e0: 00bf bf00 00bf b801 39bf 0000 bf11 0003
    00000f0: 3604 2db0 2a59 b400 1104 64b5 0011 2ab4
    0000100: 00bb b600 c1c0 013b 59b4 013d 4e59 b401
    0000110: 3e4c 59b4 0141 4d59 b401 4236 04b4 0143
    0000120: 3607 a7ff 02bb 013b 59b7 0144 592d b501
    0000130: 3d59 2bb5 013e 592c c001 0ab5 0141 5915
    0000140: 04b5 0142 5915 07b5 0143 5919 05b5 0145
    0000150: 5919 06b5 0146 3a08 2ab4 00bb 1908 b600
    0000160: fe57 2a59 b400 1104 60b5 0011 2db0     
  Exception Handler Table:
    bci [92, 223] => handler: 226
    bci [92, 223] => handler: 230
  Stackmap Table:
    full_frame(@36,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Null,Null,Integer},{})
    same_frame(@47)
    same_frame(@92)
    same_frame(@97)
    same_frame(@132)
    same_frame(@143)
    same_locals_1_stack_item_frame(@168,Object[#183])
    same_frame(@172)
    same_frame(@191)
    same_frame(@194)
    same_locals_1_stack_item_frame(@205,Object[#4])
    same_frame(@215)
    full_frame(@223,{},{Object[#264]})
    full_frame(@226,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Object[#266],Object[#266],Integer},{Object[#112]})
    full_frame(@227,{},{Object[#264]})
    full_frame(@230,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Object[#266],Object[#266],Integer},{Object[#264]})
    full_frame(@234,{},{Object[#264]})
    full_frame(@237,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Null,Null,Integer},{})
    full_frame(@244,{Object[#13],Object[#112],Null,Null,Integer,Null,Null,Integer},{})
    full_frame(@293,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Object[#266],Object[#266],Integer},{})
 
java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    issue.getPrintableStackTrace(Lio/ballerina/runtime/internal/scheduling/Strand;Lio/ballerina/runtime/internal/values/ErrorValue;)Lio/ballerina/runtime/internal/values/HandleValue; @144: invokevirtual
  Reason:
    Type 'io/ballerina/runtime/internal/values/ErrorValue' (current frame, stack[0]) is not assignable to 'io/ballerina/runtime/internal/values/HandleValue'
  Current Frame:
    bci: @144
    flags: { }
    locals: { 'io/ballerina/runtime/internal/scheduling/Strand', 'io/ballerina/runtime/internal/values/ErrorValue', 'java/lang/String', 'io/ballerina/runtime/internal/values/HandleValue', integer, null, null, integer }
    stack: { 'io/ballerina/runtime/internal/values/ErrorValue' }
  Bytecode:
    0000000: 014d 014e 0336 0401 3a05 013a 0603 3607
    0000010: 2ab4 0011 9d00 e02a 59b4 0014 5a04 60b5
    0000020: 0014 3607 2ab4 0018 9900 07b8 001e bf15
    0000030: 04ab 0000 0000 00f4 0000 0004 0000 0000
    0000040: 0000 002b 0000 0001 0000 0030 0000 0002
    0000050: 0000 00a6 0000 0003 0000 00bc 1100 0036
    0000060: 0411 0001 3604 2ab6 002c 9900 252a 03b5
    0000070: 002f 2ab4 0033 c600 0e2a b400 3359 2a01
    0000080: b500 33bf 2ab4 010d c001 0a4d a700 202b
    0000090: b601 10c0 00b7 59c7 0011 b201 15b2 011b
    00000a0: 03bd 0004 b801 21bf b601 244d 2ab6 005c
    00000b0: 9900 0f13 0126 3a05 1260 3a06 a700 69a7
    00000c0: 0003 2c59 c101 289a 0006 b801 2cb2 0130
    00000d0: b801 36c0 0106 4e11 0002 3604 a700 1100
    00000e0: 00bf bf00 00bf b801 39bf 0000 bf11 0003
    00000f0: 3604 2db0 2a59 b400 1104 64b5 0011 2ab4
    0000100: 00bb b600 c1c0 013b 59b4 013d 4e59 b401
    0000110: 3e4c 59b4 0141 4d59 b401 4236 04b4 0143
    0000120: 3607 a7ff 02bb 013b 59b7 0144 592d b501
    0000130: 3d59 2bb5 013e 592c c001 0ab5 0141 5915
    0000140: 04b5 0142 5915 07b5 0143 5919 05b5 0145
    0000150: 5919 06b5 0146 3a08 2ab4 00bb 1908 b600
    0000160: fe57 2a59 b400 1104 60b5 0011 2db0     
  Exception Handler Table:
    bci [92, 223] => handler: 226
    bci [92, 223] => handler: 230
  Stackmap Table:
    full_frame(@36,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Null,Null,Integer},{})
    same_frame(@47)
    same_frame(@92)
    same_frame(@97)
    same_frame(@132)
    same_frame(@143)
    same_locals_1_stack_item_frame(@168,Object[#183])
    same_frame(@172)
    same_frame(@191)
    same_frame(@194)
    same_locals_1_stack_item_frame(@205,Object[#4])
    same_frame(@215)
    full_frame(@223,{},{Object[#264]})
    full_frame(@226,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Object[#266],Object[#266],Integer},{Object[#112]})
    full_frame(@227,{},{Object[#264]})
    full_frame(@230,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Object[#266],Object[#266],Integer},{Object[#264]})
    full_frame(@234,{},{Object[#264]})
    full_frame(@237,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Null,Null,Integer},{})
    full_frame(@244,{Object[#13],Object[#112],Null,Null,Integer,Null,Null,Integer},{})
    full_frame(@293,{Object[#13],Object[#112],Object[#266],Object[#262],Integer,Object[#266],Object[#266],Integer},{})

        at $_init.$moduleExecute(.)
        at $_init.$lambda$$moduleExecute$(.)
        at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(SchedulerItem.java:54)
        at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:306)
        at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:273)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by gabilang**
Invalid interop calls of instance methods gave codegen errors for some cases. Handled them with this PR.

Fixes #40327

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java [hunk 1]
```diff
@@ -152,6 +152,24 @@ public void testMethodNotFound6() {
...
+        String message = "{ballerina/jballerina.java}METHOD_NOT_FOUND 'No such public static method '%s' with " +
+                "'%s' parameter(s) found in class '%s''";
+
+        BAssertUtil.validateError(compileResult, 0, String.format(message, "getPrintableStackTrace", "1",
+                        "io.ballerina.runtime.api.values.BError"), "method_not_found7.bal", 19, 1);
+
+        BAssertUtil.validateError(compileResult, 1, String.format(message, "concat", "2",
+                        "io.ballerina.runtime.api.values.BString"), "method_not_found7.bal", 23, 1);
+    }
+
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:172]:
```suggestion
    @Test
    public void testMethodNotFound7() {
        String path = "test-src/javainterop/negative/method_not_found7.bal";
        CompileResult compileResult = BCompileUtil.compile(path);
        Assert.assertEquals(compileResult.getDiagnostics().length, 2);
        String message = "{ballerina/jballerina.java}METHOD_NOT_FOUND 'No such public static method '%s' with " +
                "'%s' parameter(s) found in class '%s''";
        BAssertUtil.validateError(compileResult, 0, String.format(message, "getPrintableStackTrace", "1",
                        "io.ballerina.runtime.api.values.BError"), "method_not_found7.bal", 19, 1);
        BAssertUtil.validateError(compileResult, 1, String.format(message, "concat", "2",
                        "io.ballerina.runtime.api.values.BString"), "method_not_found7.bal", 23, 1);
    }

```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java
Comment (Nadeeshan96) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:1]: Shall we check why the lines are shown as not covered

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java [hunk 2]
```diff
@@ -60,21 +60,20 @@ public void testMethodNotFound1() {
         Assert.assertEquals(compileResult.getDiagnostics().length, 1);
         BAssertUtil.validateError(compileResult, 0,
                 "{ballerina/jballerina.java}METHOD_NOT_FOUND 'No such public method " +
-                        "'acceptStringOrErrorReturn' found in class 'class org.ballerinalang.nativeimpl." +
+                        "'acceptStringOrErrorReturn' found in class 'org.ballerinalang.nativeimpl." +
                         "jvm.tests.StaticMethods''", "method_not_found1.bal", 8, 1);
     }
 
     @Test
     public void testMethodNotFound2() {
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:69]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java [hunk 3]
```diff
@@ -83,42 +82,39 @@ public void testMethodNotFound2() {
     public void testMethodNotFound3() {
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:83]:
```suggestion
```
Shall we remove unnecessary new lines? Found in below places too

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java [hunk 4]
```diff
@@ -83,42 +82,39 @@ public void testMethodNotFound2() {
...
         BAssertUtil.validateError(compileResult, 0,
                 "{ballerina/jballerina.java}METHOD_NOT_FOUND 'No such public static method " +
                         "'acceptRecordAndRecordReturn' with '3' " + "parameter(s) found in class " +
-                        "'class org.ballerinalang.nativeimpl.jvm.tests.StaticMethods''",
+                        "'org.ballerinalang.nativeimpl.jvm.tests.StaticMethods''",
                 "method_not_found3.bal", 21, 1);
     }
 
     @Test
     public void testMethodNotFound4() {
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:97]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java [hunk 5]
```diff
@@ -83,42 +82,39 @@ public void testMethodNotFound2() {
...
         Assert.assertEquals(compileResult.getDiagnostics().length, 1);
         BAssertUtil.validateError(compileResult, 0,
                 "{ballerina/jballerina.java}METHOD_NOT_FOUND 'No such public method " +
-                        "'acceptIntAndUnionReturn' found in class 'class org.ballerinalang.nativeimpl.jvm." +
+                        "'acceptIntAndUnionReturn' found in class 'org.ballerinalang.nativeimpl.jvm." +
                         "tests.StaticMethods''", "method_not_found4.bal", 23, 1);
     }
 
     @Test
     public void testMethodNotFound5() {
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:110]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java [hunk 6]
```diff
@@ -127,36 +123,64 @@ public void testMethodNotFound6() {
...
                 testFileName, 15, 1);
-
         BAssertUtil.validateError(compileResult, 3,
                 String.format(message, "getTypeDescOnly", bTypeDescClassName + "," + bFutureClassName),
                 testFileName, 21, 1);
     }
 
+    @Test
+    public void testMethodNotFound7() {
+
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:147]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java [hunk 7]
```diff
@@ -404,6 +419,7 @@ public void testMethodSignatureNotMatch14() {
 
     @Test(description = "When there are instance and static methods with same name and parameters that differ by one")
     public void testResolveWithInstanceAndStatic() {
+
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:422]: Why have we added these?

Comment (Nadeeshan96) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:422]: Need to fix in below places too.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java (hunk 4 above)
Comment (gabilang) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:97]: Are we following any coding convention to have a new line between the method signature and method body?

Comment (Nadeeshan96) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/NegativeValidationTest.java:97]: I was told that previously there was such convention and that had been added in the checkstyle.xml file also. But that was removed and now we are following not to keep newlines between the method signature and method body.

**Review by Nadeeshan96 — APPROVED** [review]
LGTM

**Review by HindujaB — APPROVED** [review]
LGTM

**Pull request merged**